yapapi/script/__init__.py

```diff
@@ -0,0 +1,172 @@
+import asyncio
+from datetime import timedelta
+from typing import Any, Awaitable, Callable, Dict, Optional, List, Tuple, TYPE_CHECKING
+
+from yapapi.events import CommandExecuted
+from yapapi.script.capture import CaptureContext
+from yapapi.script.command import (
+    BatchCommand,
+    Command,
+    Deploy,
+    DownloadBytes,
+    DownloadFile,
+    DownloadJson,
+    Run,
+    SendBytes,
+    SendFile,
+    SendJson,
+    Start,
+    Terminate,
+)
+from yapapi.storage import DOWNLOAD_BYTES_LIMIT_DEFAULT
+
+if TYPE_CHECKING:
+    from yapapi.ctx import WorkContext
+
+
+class Script:
+    """Represents a series of commands to be executed on a provider node.
+
+    New commands are added to the script either through its `add` method or by calling one of the
+    convenience methods provided (for example: `run` or `send_json`).
+    Adding a new command *does not* result in it being immediately executed. Once ready, a `Script`
+    instance is meant to be yielded from a worker function (work generator pattern).
+    Commands will be run in the order in which they were added to the script.
+
+    If the `WorkContext` instance this `Script` uses has the field `_implicit_init` set to `True`,
+    the first script to be yielded is going to prepend the user's commands with `Deploy` and
+    `Start` commands.
+    """
+
+    timeout: Optional[timedelta] = None
+    """Time after which this script's execution should be forcefully interrupted."""
+
+    wait_for_results: bool = True
+    """Whether this script's execution should block until its results are available."""
+
+    def __init__(self, context: "WorkContext"):
+        self._ctx: "WorkContext" = context
+        self._commands: List[Tuple[Command, asyncio.Future]] = []
```

Comment:
I don't like this list of tuples. If those things are always together and never separate, maybe they should be somehow connected? Like in:
```
class Command(abc.ABC):
    def __init__(self):
        self._result = asyncio.get_event_loop().create_future()
     
     # or maybe without those methods, just use command.result?
     def set_result(self,...): self._result.set_result(...)
     def result(self): return self._result.result()
```
And just use `command.result` everywhere?

Pros:
* All futures are created in the same place, and also encapsulated nicely
* In `Engine.process_batches.get_batch_results` we'd have `return [command.result for command in script.commands]` and I think this would be much easier to understand (results here? results there? but why?)
* I don't really like the fact that we need a Script object to set the result of a command. This *could* be changed in the future, but only if we disconnect Script and results.
* We don't force the developer to keep results of `add` (they might be keeping `Command`s either way).
* It feels natural : )

I'm not sure about this, maybe there is some reason that makes this not possible.

Reply:
That's a good idea, I don't see a reason not to implement it this way. Will do!

Reply:
Implemented in c086011